### PR TITLE
Improve MIDI playback accuracy

### PR DESCRIPTION
**Pull request by Ghabry**
by processing messages in smaller chunks. This reduces the resolution from 4-8ms to 1ms.

[dk.zip](https://github.com/EasyRPG/Player/files/6137688/dk.zip)

- dkbad.ogg - Upstream
- dkgood.ogg - This patch
- dkorig.ogg - RPG_RT (different soundfont)

### CODE REVIEW

**Review by fdelapena — APPROVED** [review]

**Comment by fdelapena** [comment]
Though a previous hotfix for speed broke midi tempo in master (not just for wildmidi), https://easyrpg.org/play/pr2478/?game=smashbros note timings play properly now :+1:

**Comment by Ghabry** [comment]
The playback speed does not change anymore?
Okay will check this

**Comment by Ghabry** [comment]
My last Pitch patch had a bug: The calculation was wrong, must be 100/pitch, not pitch/100.

I also added a clamping for invalid BGM/SE values. I read in the Yume2kki Discord that a song in the latest version uses a pitch of 20. This is adjusted to 100 because it is out-of-range. Now it becomes 50. This is at least closer. Lifting the range would need testing first, so this fix was faster.

**Comment by fdelapena** [comment]
The pitch/tempo works again, thanks!
About clamping values: [RPG_RT.exe supports modified pitch values below 50% / above 150%](https://www.youtube.com/watch?v=oxT0rsHAVjo&t=78s). It's unlikely somebody modified this by hand, but based on experience with some weird games tweaking some values... well, we can wait until some bug report appears :+1:.

File: src/decoder_midigeneric.cpp [hunk 1]
```diff
@@ -25,6 +25,10 @@
 
 constexpr int GenericMidiDecoder::midi_default_tempo;
 
+// 1 ms of MIDI message resolution for a 44100 Hz samplerate
+constexpr int samples_per_play = 512;
+constexpr int bytes_per_sample = sizeof(int_least16_t) * 2;
```
Comment (fmatthew5876) [inline src/decoder_midigeneric.cpp:30]: Why `int_least16_t`? This could be larger than 16 bit.

Why not `int16_t`?

Comment (Ghabry) [reply src/decoder_midigeneric.cpp:30]: Because fmmidi used it, was a copy-paste from there :sweat_smile: . But you are right, this makes no sense.

**Review by fmatthew5876 — APPROVED** [review]

File: src/decoder_midigeneric.cpp (hunk 1 above)
Comment (Ghabry) [reply src/decoder_midigeneric.cpp]: fixed

**Pull request merged**